sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/CommentedOutCodeCheck.java

```diff
@@ -0,0 +1,132 @@
+import org.sonar.check.Rule;
+import org.sonarsource.analyzer.commons.xml.SafetyFactory;
+import org.sonarsource.analyzer.commons.xml.XmlFile;
+import org.sonarsource.analyzer.commons.xml.XmlTextRange;
+import org.sonarsource.analyzer.commons.xml.checks.SonarXmlCheck;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+@Rule(key = CommentedOutCodeCheck.RULE_KEY)
+public class CommentedOutCodeCheck extends SonarXmlCheck {
+
+  public static final String RULE_KEY = "S125";
+
+  private final XPathExpression commentsExpression = getXPathCommentsExpression();
+
+  private final Set<Node> reportedNodes = new HashSet<>();
+
+  @Override
+  protected void scanFile(XmlFile file) {
+    Charset charset = file.getInputFile().charset();
+
+    for (Node comment : getComments(file)) {
+      if (reportedNodes.contains(comment)) {
+        // already reported in previous issue
+        continue;
+      }
+      List<Node> siblingComments = getNextCommentSiblings(comment);
+      if (isParseableXml(siblingComments, charset)) {
+        reportedNodes.addAll(siblingComments);
+        XmlTextRange start = XmlFile.nodeLocation(siblingComments.get(0));
+        XmlTextRange end = XmlFile.nodeLocation(siblingComments.get(siblingComments.size() - 1));
+        reportIssue(new XmlTextRange(start, end), "Remove this commented out code.", Collections.emptyList());
+      }
+    }
+    // clear for next XML file
+    reportedNodes.clear();
+  }
+
+  private List<Node> getComments(XmlFile file) {
+    NodeList comments;
+    try {
+      comments = (NodeList) commentsExpression.evaluate(file.getDocument(), XPathConstants.NODESET);
+    } catch (XPathExpressionException e) {
+      throw new IllegalStateException("Failed to run XPath expression", e);
+    }
+    return IntStream.range(0, comments.getLength())
+      .mapToObj(comments::item)
+      .filter(comment -> comment.getTextContent().trim().startsWith("<"))
+      .collect(Collectors.toList());
+  }
+
+  private static List<Node> getNextCommentSiblings(Node comment) {
+    List<Node> results = new ArrayList<>();
+    Node current = comment;
+    while (current != null) {
+      short nodeType = current.getNodeType();
+      if (nodeType == Node.COMMENT_NODE) {
+        results.add(current);
+      } else if (nodeType != Node.TEXT_NODE || !isEmpty(current)) {
+        // any other type discard collecting siblings
+        break;
+      }
+      current = current.getNextSibling();
+    }
+    return results;
+  }
+
+  private static boolean isEmpty(Node current) {
+    return current.getTextContent().trim().isEmpty();
+  }
+
+  private static boolean isParseableXml(List<Node> siblingComments, Charset charset) {
+    String commentsAsSingleString = siblingComments.stream()
+      .map(Node::getTextContent)
+      .collect(Collectors.joining("\n"));
+
+    try (ByteArrayInputStream stream = new ByteArrayInputStream(commentsAsSingleString.getBytes(charset))) {
```

Comment:
If my understanding is correct, in this case where a random comment is in the middle of commented code:
```
<!-- <foo> -->
<!-- <bar> -->  
<!-- </bar> -->
<!-- </foo> -->
<!-- my random comment -->
<!--   <gul /> -->
<!-- <bar> -->  
<!-- </bar> -->
```
These 8 lines will be considered as "sibling" comments. Parsing will fail until the list contains only the 3 last lines. So there will be an issue only on these 3 lines.

Another case:
```
<!-- <foo> -->
<!-- <bar> -->  
<!-- </bar> -->
<!-- my random comment -->
<!-- </foo> -->
```
In this case, no issue will be raised?